### ISSUE TITLE
Avoiding Perl-ish error messages?

### ISSUE THREAD

**Issue by szabgab** [RexOps/Rex#477]
I am not sure if this is possible and if it "wanted" but let me put it here for possible discussion:

I had this file:

```
user "root";
private_key "/home/gabor/.ssh/id_rsa";
public_key "/home/gabor/.ssh/id_rsa.pub";
key_auth;

group myserver => "s7", "s8", "s9, "s11", "s14";
desc "Get the uptime of all server";
task "uptime", group => "myserver", sub {
   my $output = run "uptime";
   say $output;
};
```

which was missing a closing quote. Running `rex uptime` gave me an error about

```Bareword found where operator expected at Rexfile line 6, near ""s9, "s11"

(Missing operator before s11?)

String found where operator expected at Rexfile line 6, near "s11", ""

```

I know a bit Perl and could fix that easily but I am not sure if that Perl-ish error is really user-friendly. Would it be possible to give a better error message?
```

**Comment by ferki** [comment]
First, thanks for your patience with this issue!

## TL;DR

Rexfile loading has evolved a lot since the original report, and I consider the message handling fixed for now, because:

- only a short message, and information about failure location (file, and line number) are displayed
- debug output shows the uglier internal details
- `use diagnostics` works in Rexfiles to get verbose output from Perl itself

Rex profits a lot from the “Rex code is just Perl code” mantra, so at least some of the Perlish aspects will probably not go away completely.

Still, there’s always opportunity to do even better in even more code paths – let’s tackle them as needed in separate discussions/issues.

## Historical context

Since the original report, the Rexfile loading procedure has went through a considerable evolution, including how potential warnings and error messages are getting displayed.

For the example code and command given here, the output started out looking like this:

```
(Missing operator before s11?)

String found where operator expected at Rexfile line 6, near "s11", ""
```

After several steps, the internal approach has settled on loading the Rexfile as a Perl module to retrieve what `perl` itself thinks about it, and allow Rex to act accordingly. For a long while, the output became like this:

```
[2023-01-29 13:42:28] ERROR - Compile time errors:
[2023-01-29 13:42:28] ERROR - 	Can't find string terminator '"' anywhere before EOF at __Rexfile__.pm line 8.
[2023-01-29 13:42:28] ERROR - 	Compilation failed in require at /home/ferki/perl5/perlbrew/perls/ferki/lib/site_perl/5.36.0/Rex/CLI.pm line 759.
[2023-01-29 13:42:28] INFO - Exiting Rex...
[2023-01-29 13:42:28] INFO - Cleaning up...
```

Recently the Rexfile loading output has been made testable, the related code paths got some cleanup, and higher expectations has been set on the content itself as well. This is how it will look like from the next release:

```
[2023-01-29 13:40:14] ERROR - Compile time errors:
[2023-01-29 13:40:14] ERROR - 	Can't find string terminator '"' anywhere before EOF at Rexfile line 8.
[2023-01-29 13:40:14] INFO - Exiting Rex...
[2023-01-29 13:40:14] INFO - Cleaning up...
```

The notable highlights are:

- the messages go through the internal logging functions, and thus formatted accordingly
- info/warning/error/debug levels are handled separately and displayed accordingly (for example color coded)
- the short error message is displayed about the cause and its detected location (file, line number)
- the Rexfile path is shown correctly (as a relative path)
- the “uglier” details about the internal point of failure is only shown as a debug output

## Future ideas

### Parse error messages

> maybe https://metacpan.org/release/Parse-ErrorString-Perl [...] could be useful.

That definitely looks useful if/when someone would like to implement additional functionality, like the proposed `--check` or `--explain` flags. Hooking into the related message handling is also easier now.

### Further message improvements

While it sounds difficult to fix every historical detail at one go, newly added messages are usually already more useful. Let’s keep fixing the annoying ones.

Code around `die()` and `croak()` calls might be a good starting point for anyone looking to continue improving messages.
